.github/scripts/pytest_caching_utils.py

```diff
@@ -0,0 +1,249 @@
+    zip_file_path = temp_dir / ZIP_UPLOAD / obj_key_prefix
+
+    zip_file_path = zip_folder(cache_dir, zip_file_path)
+    obj_key = f"{obj_key_prefix}{os.path.splitext(zip_file_path)[1]}"  # Keep the new file extension
+    upload_file_to_s3(zip_file_path, bucket, obj_key)
+
+
+def download_pytest_cache(
+    pr_identifier: PRIdentifier,
+    repo: GithubRepo,
+    job_identifier: str,
+    dest_cache_dir: Path,
+    temp_dir: Path,
+    bucket: str = BUCKET,
+) -> None:
+    """
+    Downloads the pytest cache from S3. The goal is to detect any tests that have failed in the past
+    and run them first, so that the dev can get faster feedback on them.
+
+    We merge the cache from all shards since tests can get shuffled around from one shard to another
+    (based on when we last updated our stats on how long each test takes to run). This ensures that
+    even if a test moves to a different shard, that shard will know to run it first if had failed previously.
+    """
+    if not bucket:
+        bucket = BUCKET
+
+    if not isinstance(pr_identifier, PRIdentifier):
+        raise ValueError(
+            f"pr_identifier must be of type PRIdentifier, not {type(pr_identifier)}"
+        )
+
+    obj_key_prefix = _get_s3_key_prefix(pr_identifier, repo, job_identifier)
+
+    zip_download_dir = temp_dir / CACHE_ZIP_DOWNLOADS / obj_key_prefix
+
+    # downloads the cache zips for all shards
+    downloads = download_s3_objects_with_prefix(
+        bucket, obj_key_prefix, zip_download_dir
+    )
+
+    for downloaded_zip in downloads:
+        # the file name of the zip is the shard id
+        shard = os.path.splitext(os.path.basename(downloaded_zip))[0]
+        cache_dir_for_shard = _get_temp_cache_dir_path(
+            temp_dir, pr_identifier, repo, job_identifier, shard
+        )
+
+        unzip_folder(downloaded_zip, cache_dir_for_shard)
+        print(
+            f"Merging cache for job_identifier `{job_identifier}`, shard `{shard}` into `{dest_cache_dir}`"
+        )
+        _merge_pytest_caches(cache_dir_for_shard, dest_cache_dir)
+
+
+def _get_temp_cache_dir_path(
+    temp_dir: Path,
+    pr_identifier: PRIdentifier,
+    repo: GithubRepo,
+    job_identifier: str,
+    shard: str,
+) -> Path:
+    return (
+        temp_dir
+        / UNZIPPED_CACHES
+        / _get_s3_key_prefix(pr_identifier, repo, job_identifier, shard)
+        / PYTEST_CACHE_DIR_NAME
+    )
+
+
+def _get_s3_key_prefix(
+    pr_identifier: PRIdentifier,
+    repo: GithubRepo,
+    job_identifier: str,
+    shard: str = "",
+) -> str:
+    """
+    The prefix to any S3 object key for a pytest cache. It's only a prefix though, not a full path to an object.
+    For example, it won't include the file extension.
+    """
+    prefix = f"{PYTEST_CACHE_KEY_PREFIX}/{repo.owner}/{repo.name}/{pr_identifier}/{sanitize_for_s3(job_identifier)}"
+
+    if shard:
+        prefix += f"/{shard}"
+
+    return prefix
+
+
+def _merge_pytest_caches(
```

Comment:
cc @clee2000 Do you think that this would work with the new stepcurrent plugin https://github.com/pytorch/pytorch/blob/main/test/conftest.py#L21?  As this plugin saves the last test runs into the cache and resumes from that, I have a feeling that we don't need to persist this information across jobs (at least not now)

AFAIK, there are at least the following entries in pytest cache:
* stepcurrent (or its official cousin stepwise)
* lastfailed

Do we need to persist anything from the cache besides `lastfailed`?

Reply:
as of right now, the only file that needs to be saved is lastfailed.  In the future nodeids might also be useful for running new tests first as well.  Stepcurrent and stepwise related files and folders should definitely not be copied

Reply:
Thanks for the context! Limiting caching to just the static support files (which pytest [doesn't recreate](https://github.com/pytest-dev/pytest/blob/62320e4ff77af1e50cc93cd664a7549ae08338b7/src/_pytest/cacheprovider.py#L185-L186C31) if the .pytest_cache dir already exists) and the lastfailed file.